### PR TITLE
Lock file during download

### PR DESCRIPTION
**Pull request by PVince81**
Put a shared lock on a file before download it, and free it at the end of the transfer.

My worries here:
1) Shared lock during download will prevent other users to update/overwrite that file in the time being. Ideal would be able to deliver an older copy of the file
2) What happens if the download takes too long and PHP times out, will the lock stay forever ?

@DeepDiver1975 @icewind1991 what do you think ?

A similar approach will be needed for zip files, see https://github.com/owncloud/core/issues/16960#issuecomment-112482247

### CODE REVIEW

**Comment by PVince81** [comment]
This for the download error message part for here https://github.com/owncloud/core/issues/16701

Another alternative would be to just try and lock the file before the upload, and then directly unlock before the transfer starts, or after the first packet. (hacky)

**Comment by PVince81** [comment]
Looks like max execution time is set to 0 here https://github.com/owncloud/core/blob/master/lib/private/files.php#L151 but it's only for zip files, not regular downloads.

**Comment by DeepDiver1975** [comment]
> 1) Shared lock during download will prevent other users to update/overwrite that file in the time being. Ideal would be able to deliver an older copy of the file

well - this is what we want to have - right? Don't touch the file while I'm downloading it or I'll get crap.



> 2) What happens if the download takes too long and PHP times out, will the lock stay forever ?

This is why we need a TTL on locks - as discussed right from the beginning

**Comment by DeepDiver1975** [comment]
> Looks like max execution time is set to 0 here https://github.com/owncloud/core/blob/master/lib/private/files.php#L151 but it's only for zip files, not regular downloads.

we might want to change this and have execution time of 0 for any kind of download.
(Note: this whole messy file will die in 8.2 due to webdav :tada: )

File: lib/private/files.php [hunk 1]
```diff
@@ -180,6 +184,13 @@ public static function get($dir, $files, $only_header = false) {
 					\OC\Files\Filesystem::readfile($filename);
 				}
 			}
+			if ($get_type === self::FILE) {
+				$view->unlockFile($filename, ILockingProvider::LOCK_SHARED);
+			}
+		} catch (\OCP\Lock\LockedException $ex) {
+			$l = \OC::$server->getL10N('core');
+			$hint = method_exists($ex, 'getHint') ? $ex->getHint() : '';
+			\OC_Template::printErrorPage($l->t('File is currently busy, please try again later'), $hint);
```
Comment (DeepDiver1975) [inline lib/private/files.php]: brrr - this file has to die :skull:

Comment (PVince81) [inline lib/private/files.php]: We still need it for zip download... unless we can come up with a WebDAV endpoint for zip download.

Also see https://github.com/owncloud/core/issues/16960#issuecomment-113102628, very hard to decide whether to move forward

**Comment by PVince81** [comment]
TTL discussion here: https://github.com/owncloud/core/issues/16966

File: lib/private/files.php (hunk 1 above)
Comment (DeepDiver1975) [inline lib/private/files.php]: > We still need it for zip download... unless we can come up with a WebDAV endpoint for zip download.

we have to do this anyhow from my pov

Comment (PVince81) [inline lib/private/files.php]: Raised zip webdav endpoint here https://github.com/owncloud/core/issues/17014

**Comment by icewind1991** [comment]
Shouldn't the fopen already handle the locking here?

**Comment by PVince81** [comment]
Yes, it does. But then it's too late and we already sent the 200 OK + headers.

We need to be able to return another HTTP code before fopen (readfile) happens.

**Comment by DeepDiver1975** [comment]
Should we simply disable xsendfile as soon as locking is enabled

**Comment by DeepDiver1975** [comment]
> Should we simply disable xsendfile as soon as locking is enabled?

or drop this feature completely?

**Comment by DeepDiver1975** [comment]
> or drop this feature completely?

especially thinking about all our wrapper setups I really question if there is a single valid scenario for xsendfile left.

**Comment by karlitschek** [comment]
lets disable it for now if filelocking is active and discuss again for 8.2

**Comment by DeepDiver1975** [comment]
> lets disable it for now if filelocking is active and discuss again for 8.2

agreed - I'll add this to this pr ...

**Comment by PVince81** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
- [ ] disable xsendfile if locking is enabled

**Comment by DeepDiver1975** [comment]
> disable xsendfile if locking is enabled

@PVince81 c74c8ef

**Comment by DeepDiver1975** [comment]
Jenkins fails on oracle - can be ignored :+1:

**Comment by DeepDiver1975** [comment]
@th3fallen  @icewind1991 please review - thx

**Comment by icewind1991** [comment]
:+1: looks good

**Pull request merged**